views.go

```diff
@@ -0,0 +1,221 @@
+package slack
+
+import (
+	"context"
+	"encoding/json"
+)
+
+const (
+	VTModal   ViewType = "modal"
+	VTHomeTab ViewType = "home"
+)
+
+type ViewType string
+
+type View struct {
+	SlackResponse
+	ID              string           `json:"id"`
+	TeamID          string           `json:"team_id"`
+	Type            ViewType         `json:"type"`
+	Title           *TextBlockObject `json:"title"`
+	Close           *TextBlockObject `json:"close"`
+	Submit          *TextBlockObject `json:"submit"`
+	Blocks          Blocks           `json:"blocks"`
+	PrivateMetadata string           `json:"private_metadata"`
+	CallbackID      string           `json:"callback_id"`
+	State           interface{}      `json:"state"`
+	Hash            string           `json:"hash"`
+	ClearOnClose    bool             `json:"clear_on_close"`
+	NotifyOnClose   bool             `json:"notify_on_close"`
+	RootViewID      string           `json:"root_view_id"`
+	PreviousViewID  string           `json:"previous_view_id"`
+	AppID           string           `json:"app_id"`
+	ExternalID      string           `json:"external_id"`
+	BotID           string           `json:"bot_id"`
+}
+
+type ModalViewRequest struct {
+	Type            ViewType         `json:"type"`
+	Title           *TextBlockObject `json:"title"`
+	Blocks          Blocks           `json:"blocks"`
+	Close           *TextBlockObject `json:"close"`
+	Submit          *TextBlockObject `json:"submit"`
+	PrivateMetadata string           `json:"private_metadata"`
+	CallbackID      string           `json:"callback_id"`
+	ClearOnClose    bool             `json:"clear_on_close"`
+	NotifyOnClose   bool             `json:"notify_on_close"`
+	ExternalID      string           `json:"external_id"`
+}
+
+func (v *ModalViewRequest) ViewType() ViewType {
+	return v.Type
+}
+
+type HomeTabViewRequest struct {
+	Type            ViewType `json:"type"`
+	Blocks          Blocks   `json:"blocks"`
+	PrivateMetadata string   `json:"private_metadata"`
+	CallbackID      string   `json:"callback_id"`
+	ExternalID      string   `json:"external_id"`
+}
+
+func (v *HomeTabViewRequest) ViewType() ViewType {
+	return v.Type
+}
+
+type openViewRequest struct {
+	TriggerID string           `json:"trigger_id"`
+	View      ModalViewRequest `json:"view"`
+}
+
+type publishViewRequest struct {
+	UserID string             `json:"user_id"`
+	View   HomeTabViewRequest `json:"view"`
+	Hash   string             `json:"hash"`
```

Comment:
@kanata2 I've been trying this out and I came across an issue due to `hash` being optional https://api.slack.com/methods/views.publish . If you supply an empty string as the hash value (in order to not send the hash value) the API request still includes the empty string, this causes Slack to give the `hash_conflict` error if you are trying to publish a view that already has a hash defined. The fix is easy though! Omit the attribute from the API request payload if it's empty. This will also apply to the `ViewUpdate` method too.
```
Hash   string             `json:"hash,omitempty"`
```

Reply:
Thank you for your review! I missed it...
Handle optional values as `omitempty` in #c14171d